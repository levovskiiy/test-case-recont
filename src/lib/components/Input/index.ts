import Input from './src/Input.vue'

export { Input }
export default Input
export * from './src/input'
