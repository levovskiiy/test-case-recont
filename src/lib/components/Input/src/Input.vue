<script lang="ts" setup>
import { computed, ref, useAttrs, useTemplateRef } from 'vue'
import { useAttrs as useExcludedAttrs } from '@/lib/composables'
import { inputEmits, inputProps } from './input'

defineOptions({
  inheritAttrs: false,
})

const props = defineProps(inputProps)
const emit = defineEmits(inputEmits)

const focused = ref(false)
const inputRef = useTemplateRef('input')

const allAttrs = useAttrs()
const inputAttrs = useExcludedAttrs()
const containerCls = computed(() => [
  {
    focused: focused.value,
    error: props.error,
  },
  allAttrs.class,
])

function onFocus(e: FocusEvent) {
  focused.value = true
  emit('focus', e)
}

function onBlur(e: FocusEvent) {
  focused.value = false
  emit('blur', e)
}

function onInput(event: Event) {
  const { value } = event.target as HTMLInputElement
  emit('update:model-value', value)
}
</script>

<template>
  <div class="input">
    <slot name="left" />
    <div class="input-wrapper" @click.stop="inputRef?.focus()" :class="containerCls">
      <input
        ref="input"
        v-bind="inputAttrs"
        :value="modelValue"
        :placeholder="placehodler"
        class="input-field"
        @focus="onFocus"
        @blur="onBlur"
        @input="onInput"
      />
    </div>
    <slot name="right" />
  </div>
</template>

<style lang="scss" scoped>
.input {
  --default-color: 28, 125, 65;
  --error-color: 144, 11, 9;

  min-height: 40px;
  box-sizing: border-box;
  position: relative;
  display: inline-flex;
  width: 100%;
  vertical-align: middle;

  .input-wrapper {
    padding: 10px 12px;
    border-radius: 8px;
    display: inline-flex;
    flex: 1;
    align-items: center;
    justify-content: center;
    width: 100%;
    transition: all 0.2s;
    background: rgba(245, 246, 246, 1);

    &.focused {
      background: rgba(var(--default-color), 0.05);
      box-shadow: 0 0 0 2px rgb(var(--default-color)) inset;
    }

    &.error {
      background: rgba(var(--error-color), 0.05);
      box-shadow: 0 0 0 2px rgb(var(--error-color)) inset;
    }

    .input-field {
      box-sizing: border-box;
      width: 100%;
      flex-grow: 1;
      padding: 0;
      border: none;
      outline: none;
      background: none;
      box-shadow: none;

      &::placeholder {
        color: rgba(112, 119, 134, 1);
      }
    }
  }
}
</style>
