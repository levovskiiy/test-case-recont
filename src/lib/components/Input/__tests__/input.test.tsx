import Input from '../src/Input.vue'
import { mount } from '@vue/test-utils'
import { describe, test, expect, vi } from 'vitest'
import { nextTick, ref } from 'vue'

describe('input', () => {
  test('Обновляет переданное состояние ввода', async () => {
    const value = ref('test')

    const wrapper = mount(() => <Input v-model={value.value} />)

    const input = wrapper.find('input')
    expect(input.element.value).toMatchSnapshot('test')

    value.value = '1'
    await nextTick()
    expect(input.element.value).toMatchSnapshot('1')

    wrapper.vm.$emit('update:model-value', '123')
    await nextTick()
    expect(value.value).toMatchSnapshot('123')
  })

  describe('Input events', () => {
    const focus = vi.fn()
    const blur = vi.fn()

    test('focus event', async () => {
      const wrapper = mount(() => <Input onFocus={focus} />)

      const input = wrapper.find('input')
      await input.trigger('focus')
      expect(focus).toHaveBeenCalled()
    })

    test('blur event', async () => {
      const wrapper = mount(() => <Input onBlur={blur} />)

      const input = wrapper.find('input')
      await input.trigger('blur')
      expect(blur).toHaveBeenCalled()
    })
  })
})
