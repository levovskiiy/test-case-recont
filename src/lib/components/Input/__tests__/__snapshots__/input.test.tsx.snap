// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`input > Обновляет переданное состояние ввода > 1 1`] = `"1"`;

exports[`input > Обновляет переданное состояние ввода > 123 1`] = `"1"`;

exports[`input > Обновляет переданное состояние ввода > test 1`] = `"test"`;
